### ISSUE TITLE
[BUG] Factory functions declared noexcept do not work with py::init()

### ISSUE THREAD

**Issue by dkavolis** [pybind/pybind11#2666]
## Issue description

A modified example from the docs with `noexcept` factory function will not compile.

```
error C2027: use of undefined type 'pybind11::detail::initimpl::factory<Func,pybind11::detail::void_type (__cdecl *)(void),init::Example (int) noexcept,pybind11::detail::void_type (void)>'
        with
        [
            Func=init::Example (__cdecl *)(int) noexcept
        ]
```

## Reproducible example code

```cpp
class Example {
   private:
    Example(int) noexcept; // private constructor
   public:
    // Factory function:
    static Example create(int a) noexcept { return Example(a); }
  };

  py::class_<Example>(m, "Example")
      .def(py::init(&Example::create));
```

**Comment by bstaletic** [comment]
I can't repro with gcc or clang. Does `py::init([](int a) { return Example::create(a); })` work?

**Comment by dkavolis** [comment]
I have only tried MSVC. And yes, either removing `noexcept` from the factory function or passing a lambda without `noexcept` works. Adding `noexcept` to either fails to compile. It's as if MSVC treats `noexcept` as part of the type.

**Comment by ikicic** [comment]
Could it be related to C++17? Compilation fails for me when using C++17 (g++ 7.3), but works when I either remove `noexcept` or switch to C++14.

If I understand correctly, C++17 introduced pointers to `noexcept` functions, which are not covered by pybind's type traits (e.g. see `remove_class` [here](https://github.com/pybind/pybind11/blob/master/include/pybind11/detail/common.h#L545-L547))

**Comment by dkavolis** [comment]
That must be it, it compiles with C++14 in MSVC as well.

**Comment by bstaletic** [comment]
Right. I don't know what I was testing, but I can confirm this issue. Anyway, [this thing](https://github.com/pybind/pybind11/blob/master/include/pybind11/detail/init.h#L219) needs to be fixed. The paper that introduced this change in C++17 was [P0014](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2015/p0014r1.html)

**Comment by georgthegreat** [comment]
I can confirm the issue.

At the time python bindings for [NVIDIA/TensorRT](https://github.com/NVIDIA/TensorRT/tree/release/8.6/python) can not be compiler in post-c++17 mode.
